### PR TITLE
Make python module more robust

### PR DESCRIPTION
**Pull request by SoapGentoo**
* Lookup is now performed according to the following order:
  1. use `pkg-config` with `PKG_CONFIG_LIBDIR=LIBPC` from python
  2. use plain `PKG_CONFIG_{LIBDIR,PATH}` from the environment
  3. try to extract the information from SYSCONFIG

The second step is necessary for relocated python installations
and cross compilation scenarios, where the value of `LIBPC` will
be wrong.

#### standard installation:
```
Found "python-2.7" via pkg-config lookup in LIBPC (/usr/lib64/pkgconfig)
```

#### If installation has been relocated:
```
"python-2.7" cannot be found in LIBPC (BROKEN/usr/lib64/pkgconfig), this is likely due to a relocated python installation
Found "python-2.7" via fallback pkg-config lookup in PKG_CONFIG_LIBDIR/PKG_CONFIG_PATH
```

#### if installation has been relocated AND pc files are hosed:
```
"python-2.7" cannot be found in LIBPC (BROKEN/usr/lib64/pkgconfig), this is likely due to a relocated python installation
"python-2.7" cannot be found via fallback pkg-config lookup in PKG_CONFIG_LIBDIR/PKG_CONFIG_PATH
Found "python-2.7" via SYSCONFIG module
```

### CODE REVIEW

**Comment by MathieuDuponchelle** [comment]
Excellent, the addition of debug will also definitely be useful, sorry about that. The patch is accepted for me, as long as CI is OK. I would rather you didn't number your comments, but that's not my codebase so I won't request that (but please amend this :P)

File: mesonbuild/modules/python.py [hunk 1]
```diff
@@ -64,29 +64,46 @@ def __init__(self, python_holder, environment, kwargs):
...
-                pass
+                if pkg_libdir:
+                    os.environ['PKG_CONFIG_LIBDIR'] = pkg_libdir
 
-            if old_pkg_path is not None:
-                os.environ['PKG_CONFIG_PATH'] = old_pkg_path
+                try:
+                    self.pkgdep = PkgConfigDependency('python-{}'.format(pkg_version), environment, kwargs)
+                    mlog.debug('Found "python-{}" via pkg-config lookup in LIBPC ({})'.format(pkg_version, pkg_libdir))
+                except Exception:
```
Comment (nirbheek) [inline mesonbuild/modules/python.py]: This should be `MesonException`. The original code used `Exception`, but that's wrong too.

Also we should log the exception using mlog.debug.

File: mesonbuild/modules/python.py [hunk 2]
```diff
@@ -64,29 +64,46 @@ def __init__(self, python_holder, environment, kwargs):
...
-                os.environ.pop('PKG_CONFIG_LIBDIR', None)
+                mlog.debug('"python-{}" cannot be found in LIBPC ({}), this is likely due to a relocated python installation'.format(pkg_version, pkg_libdir))
+
+            # b. If lookup via LIBPC failed, try to use fallback
+            #    PKG_CONFIG_LIBDIR/PKG_CONFIG_PATH mechanisms
+            if self.pkgdep is None or not self.pkgdep.found():
+                try:
+                    self.pkgdep = PkgConfigDependency('python-{}'.format(pkg_version), environment, kwargs)
+                    mlog.debug('Found "python-{}" via fallback pkg-config lookup in PKG_CONFIG_LIBDIR/PKG_CONFIG_PATH'.format(pkg_version))
+                except Exception:
```
Comment (nirbheek) [inline mesonbuild/modules/python.py]: Same here, `MesonException`.

File: mesonbuild/modules/python.py [hunk 3]
```diff
@@ -96,12 +113,17 @@ def __init__(self, python_holder, environment, kwargs):
...
                 if mesonlib.is_windows():
                     self._find_libpy_windows(environment)
                 else:
                     self._find_libpy(python_holder, environment)
 
+                if self.is_found:
+                    mlog.debug('Found "python-{}" via SYSCONFIG module'.format(self.version))
+
         if self.is_found:
             mlog.log('Dependency', mlog.bold(self.name), 'found:', mlog.green('YES'))
```
Comment (nirbheek) [inline mesonbuild/modules/python.py]: We should print here what mechanism was used to find the dependency, `pkgconfig` vs `sysconfig`.

**Pull request merged**